File: frontend/src/modules/forecasting-scheduling/workforce-analytics/queue-analytics.js
/* Workforce Analytics > Queue Analytics — read-only dashboard assembling
   existing telemetry for one org unit ("queue"): queue profile, service
   level target, forecast accuracy, latest backlog reading, and campaign
   membership. Wired to forecasting-service. No real-time occupancy/AHT/
   abandon-rate here — the backend deliberately doesn't invent live
   telephony signals it has no source for (see queue_analytics.py's own
   doc comment); this screen only shows what that endpoint actually
   returns. */
import { esc, errMsg } from '../../../core/dom.js';
import { doRerender } from '../../../app/rerender.js';
import { pageHead, sec, empty, fmtDt } from '../../identity-org/shared/ui.js';
import { loadOrgUnits } from '../../identity-org/shared/loaders.js';
import * as ForecastingApi from '../../forecasting/api.js';

const isOk = (x) => Array.isArray(x);
const pct = (fraction) => (fraction == null ? "—" : `${(Number(fraction) * 100).toFixed(1)}%`);

function backlogBadge(status) {
  const map = { ok: ["badge-ok", "OK"], warning: ["badge-warn", "Warning"], critical: ["badge-danger", "Critical"], unknown: ["badge-sys", "No template"] };
  const [c, l] = map[status] || ["badge-sys", status];
  return `<span class="badge ${c}"><span class="pip"></span>${l}</span>`;
}

function campaignBadge(status) {
  const map = { draft: ["badge-sys", "Draft"], active: ["badge-ok", "Active"], paused: ["badge-warn", "Paused"], completed: ["badge-off", "Completed"] };
  const [c, l] = map[status] || ["badge-sys", status];
  return `<span class="badge ${c}"><span class="pip"></span>${l}</span>`;
}

async function loadAnalytics(state, orgUnitId) {
  state.wf.queueAnalytics = state.wf.queueAnalytics || {};
  try {
    state.wf.queueAnalytics[orgUnitId] = await ForecastingApi.getQueueAnalytics(orgUnitId);
  } catch (err) {
    state.wf.queueAnalytics[orgUnitId] = { error: errMsg(err) };
  }
  doRerender();
}

export function render(state) {
  if (!isOk(state.data.orgUnits)) loadOrgUnits(state);
  const orgUnits = isOk(state.data.orgUnits) ? state.data.orgUnits : [];
  if (state.qaOu === undefined && orgUnits.length) state.qaOu = orgUnits[0].id;

  if (!orgUnits.length) {
    return `${pageHead("Queue Analytics", "Read-only rollup of forecast accuracy, SLA targets, backlog, and campaigns for one queue — wired to forecasting-service.", "")}
      <div class="panel">${empty("No org units yet.")}</div>`;
  }

  state.wf.queueAnalytics = state.wf.queueAnalytics || {};
  const a = state.wf.queueAnalytics[state.qaOu];
  if (a === undefined) loadAnalytics(state, state.qaOu);

  let body = `<div class="skel" style="height:24px"></div>`;
  if (a && a.error) body = `<p class="muted">${esc(a.error)}</p>`;
  else if (a) {
    const qp = a.queueProfile;
    const slt = a.serviceLevelTarget;
    const acc = a.accuracy;
    body = `
      ${sec("Queue profile", !qp ? `<p class="muted">Not configured for this org unit.</p>` : `<dl class="kv">
        <dt>Industry</dt><dd>${esc(qp.industry ?? "—")}</dd>
        <dt>Queue type</dt><dd>${esc(qp.queueType ?? "—")}</dd>
        <dt>Expected volume band</dt><dd>${esc(qp.expectedVolumeBand ?? "—")}</dd>
        <dt>Timezone bucket</dt><dd>${esc(qp.timezoneBucket ?? "—")}</dd>
      </dl>`)}
      ${sec("Service level target", `<dl class="kv">
        <dt>Target service level</dt><dd>${pct(slt.targetServiceLevel)}</dd>
        <dt>Target answer time</dt><dd>${slt.targetAnswerTimeSeconds}s</dd>
        <dt>Max occupancy</dt><dd>${pct(slt.maxOccupancy)}</dd>
      </dl>${slt.isDefault ? `<p class="hint">Platform default — not configured for this org unit. See Forecasting &gt; Goals &amp; Requirements.</p>` : ""}`)}
      ${sec("Forecast accuracy", `<dl class="kv">
        <dt>Accuracy points</dt><dd>${acc.pointCount}</dd>
        <dt>Avg MAPE</dt><dd>${acc.avgMape != null ? `${Number(acc.avgMape).toFixed(1)}%` : "—"}</dd>
        <dt>Latest evaluated</dt><dd>${acc.latestEvaluatedAt ? fmtDt(acc.latestEvaluatedAt) : "—"}</dd>
        <dt>Latest actual vs predicted volume</dt><dd>${acc.latestActualVolume != null ? `${acc.latestActualVolume} vs ${acc.latestPredictedVolume ?? "—"}` : "—"}</dd>
      </dl>`)}
      ${sec("Latest backlog", !a.latestBacklog ? `<p class="muted">No backlog snapshot recorded yet. See Forecasting &gt; Backlog Age.</p>` : `<dl class="kv">
        <dt>Item count</dt><dd>${a.latestBacklog.itemCount}</dd>
        <dt>Oldest item age</dt><dd>${a.latestBacklog.oldestItemAgeMinutes} min</dd>
        <dt>Recorded</dt><dd>${fmtDt(a.latestBacklog.recordedAt)}</dd>
        <dt>Status</dt><dd>${backlogBadge(a.latestBacklog.status)}</dd>
      </dl>`)}
      ${sec("Campaigns", !a.campaigns.length ? `<p class="muted">Not assigned to any campaign's queues. See Campaigns &gt; Queues.</p>` : a.campaigns.map((c) => `<div class="cov"><div><b>${esc(c.name)}</b></div><div>${campaignBadge(c.status)}</div></div>`).join(""))}
    `;
  }

  return `
    ${pageHead("Queue Analytics", "Read-only rollup of forecast accuracy, SLA targets, backlog, and campaigns for one queue — wired to forecasting-service.", "")}
    <div class="field" style="max-width:360px;margin-bottom:14px">
      <label>Org unit (queue)</label>
      <select data-wf="qa-ou">${orgUnits.map((o) => `<option value="${o.id}" ${o.id === state.qaOu ? "selected" : ""}>${"— ".repeat(o.depth)}${esc(o.name)}</option>`).join("")}</select>
    </div>
    ${body}`;
}

export function handle(state, act, id, value) {
  if (act === "qa-ou") { state.qaOu = value; return true; }
  return false;
}

File: forecasting-service/app/api/v1/queue_analytics.py
""""Queue Analytics" under Workforce Analytics in the reference console's
own menu - a read-only aggregation of signals this platform already has for
a queue (org unit): forecast accuracy, service-level-target adherence,
latest Backlog Age reading, queue profile, and campaign membership. No new
domain model - `QueueProfile`'s own doc comment already establishes org
units as queues here, so this assembles existing telemetry rather than
inventing real-time occupancy/AHT/abandon-rate metrics this platform has no
live telephony signal for.
"""

from __future__ import annotations

import uuid
from decimal import Decimal

from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db_session, get_tenant_context
from app.api.v1.schemas import (
    QueueAnalyticsAccuracySummary,
    QueueAnalyticsBacklog,
    QueueAnalyticsCampaign,
    QueueAnalyticsResponse,
    QueueProfileResponse,
    ServiceLevelTargetResponse,
)
from app.core.tenant_context import TenantContext
from app.db.models import BacklogAgeTemplate, Campaign, CampaignQueue, QueueProfile, ServiceLevelTarget
from app.services import accuracy_service, backlog_snapshot_service, headcount_service

router = APIRouter(prefix="/v1/forecasting/queue-analytics", tags=["forecasting-queue-analytics"])


def _backlog_status(age_minutes: int, template: BacklogAgeTemplate | None) -> str:
    if template is None:
        return "unknown"
    if age_minutes >= template.critical_threshold_minutes:
        return "critical"
    if age_minutes >= template.warning_threshold_minutes:
        return "warning"
    return "ok"


@router.get("/{org_unit_id}", response_model=QueueAnalyticsResponse)
async def get_queue_analytics(
    org_unit_id: uuid.UUID,
    context: TenantContext = Depends(get_tenant_context),
    session: AsyncSession = Depends(get_db_session),
) -> QueueAnalyticsResponse:
    tenant_id = uuid.UUID(context.tenant_id)

    queue_profile_row = await session.scalar(
        select(QueueProfile).where(
            QueueProfile.tenant_id == tenant_id, QueueProfile.org_unit_id == org_unit_id
        )
    )
    queue_profile = (
        QueueProfileResponse(
            org_unit_id=org_unit_id,
            industry=queue_profile_row.industry,
            queue_type=queue_profile_row.queue_type,
            expected_volume_band=queue_profile_row.expected_volume_band,
            timezone_bucket=queue_profile_row.timezone_bucket,
        )
        if queue_profile_row
        else None
    )

    slt_row = await session.scalar(
        select(ServiceLevelTarget).where(
            ServiceLevelTarget.tenant_id == tenant_id, ServiceLevelTarget.org_unit_id == org_unit_id
        )
    )
    service_level_target = (
        ServiceLevelTargetResponse(
            org_unit_id=org_unit_id,
            target_service_level=slt_row.target_service_level,
            target_answer_time_seconds=slt_row.target_answer_time_seconds,
            max_occupancy=slt_row.max_occupancy,
            is_default=False,
        )
        if slt_row
        else ServiceLevelTargetResponse(
            org_unit_id=org_unit_id,
            target_service_level=Decimal(str(headcount_service.DEFAULT_TARGET_SERVICE_LEVEL)),
            target_answer_time_seconds=headcount_service.DEFAULT_TARGET_ANSWER_TIME_SECONDS,
            max_occupancy=Decimal(str(headcount_service.DEFAULT_MAX_OCCUPANCY)),
            is_default=True,
        )
    )

    accuracy_rows = await accuracy_service.get_accuracy_trend(
        session, tenant_id=tenant_id, org_unit_id=org_unit_id
    )
    mapes = [row.mape for row in accuracy_rows if row.mape is not None]
    latest_accuracy = accuracy_rows[-1] if accuracy_rows else None
    accuracy = QueueAnalyticsAccuracySummary(
        point_count=len(accuracy_rows),
        avg_mape=(sum(mapes, Decimal(0)) / len(mapes)) if mapes else None,
        latest_evaluated_at=latest_accuracy.evaluated_at if latest_accuracy else None,
        latest_actual_volume=latest_accuracy.actual_volume if latest_accuracy else None,
        latest_predicted_volume=latest_accuracy.predicted_volume if latest_accuracy else None,
    )

    backlog_rows = await backlog_snapshot_service.list_backlog_snapshots(
        session, tenant_id=tenant_id, org_unit_id=org_unit_id, limit=1
    )
    latest_backlog = None
    if backlog_rows:
        latest = backlog_rows[0]
        template = (
            await session.scalar(
                select(BacklogAgeTemplate).where(
                    BacklogAgeTemplate.tenant_id == tenant_id, BacklogAgeTemplate.id == latest.template_id
                )
            )
            if latest.template_id
            else None
        )
        latest_backlog = QueueAnalyticsBacklog(
            item_count=latest.item_count,
            oldest_item_age_minutes=latest.oldest_item_age_minutes,
            recorded_at=latest.recorded_at,
            status=_backlog_status(latest.oldest_item_age_minutes, template),
        )

    campaign_rows = await session.execute(
        select(Campaign)
        .join(CampaignQueue, CampaignQueue.campaign_id == Campaign.id)
        .where(CampaignQueue.tenant_id == tenant_id, CampaignQueue.org_unit_id == org_unit_id)
    )
    campaigns = [
        QueueAnalyticsCampaign(id=row.id, name=row.name, status=row.status)
        for row in campaign_rows.scalars().all()
    ]

    return QueueAnalyticsResponse(
        org_unit_id=org_unit_id,
        queue_profile=queue_profile,
        service_level_target=service_level_target,
        accuracy=accuracy,
        latest_backlog=latest_backlog,
        campaigns=campaigns,
    )
